kobo-uncaged/device/device.go

```diff
@@ -584,153 +610,73 @@ func (k *Kobo) SaveCoverImage(contentID string, size image.Point, imgB64 string)
-		}
+	if len(k.UpdatedMetadata) == 0 {
+		return nil
 	}
-	// Update statment for books already in the content table
-	updateQuery := `
-		UPDATE content SET 
-		Description=?,
-		Series=?,
-		SeriesNumber=?,
-		SeriesNumberFloat=?,
-		SeriesID=?
-		WHERE ContentID=?;`
-	updateStmt, err := tx.Prepare(updateQuery)
+	updateSQL, err := newSQLWriter(filepath.Join(k.BKRootDir, kuUpdatedSQL))
 	if err != nil {
-		tx.Rollback()
-		return rerun, fmt.Errorf("UpdateNickelDB: prepared statement failed: %w", err)
+		return fmt.Errorf("WriteUpdatedMetadataSQL: failed to create SQL writer: %w", err)
 	}
+	defer updateSQL.close()
+	dialect := goqu.Dialect("sqlite3")
+	var desc, series, seriesNum *string
+	var seriesNumFloat *float64
 	for cid := range k.UpdatedMetadata {
-		desc, series, seriesID, seriesNum, seriesNumFloat = nil, nil, nil, nil, nil
+		desc, series, seriesNum, seriesNumFloat = nil, nil, nil, nil
 		if k.MetadataMap[cid].Comments != nil && *k.MetadataMap[cid].Comments != "" {
 			desc = k.MetadataMap[cid].Comments
 		}
 		if k.MetadataMap[cid].Series != nil && *k.MetadataMap[cid].Series != "" {
 			// TODO: Fuzzy series matching to deal with 'The' prefixes and 'Series' postfixes?
 			series = k.MetadataMap[cid].Series
-			seriesID = series
-			if sID, ok := k.SeriesIDMap[*series]; ok {
-				seriesID = &sID
-			}
 		}
 		if k.MetadataMap[cid].SeriesIndex != nil && *k.MetadataMap[cid].SeriesIndex != 0.0 {
 			sn := strconv.FormatFloat(*k.MetadataMap[cid].SeriesIndex, 'f', -1, 64)
 			seriesNum = &sn
 			seriesNumFloat = k.MetadataMap[cid].SeriesIndex
 		}
-		// We rollback on any sort of error, to lessen any chance of database corruption
-		if _, ok := k.BooksInDB[cid]; ok {
-			_, err = updateStmt.Exec(desc, series, seriesNum, seriesNumFloat, seriesID, cid)
-			if err != nil {
-				tx.Rollback()
-				return rerun, fmt.Errorf("UpdateNickelDB: %w", err)
-			}
-			delete(k.UpdatedMetadata, cid)
-		} else {
-			rerun = true
-			if k.KuConfig.AddMetadataByTrigger {
-				_, err = insertStmt.Exec(cid, desc, series, seriesNum, seriesNumFloat, seriesID)
-				if err != nil {
-					tx.Rollback()
-					return rerun, fmt.Errorf("UpdateNickelDB: %w", err)
-				}
-				delete(k.UpdatedMetadata, cid)
-			}
+		ds := dialect.Update("content").Set(goqu.Record{
+			"Description": desc, "Series": series, "SeriesNumber": seriesNum, "SeriesNumberFloat": seriesNumFloat,
+		}).Where(goqu.Ex{"ContentID": cid})
+		sqlStr, _, err := ds.ToSQL()
+		if err != nil {
+			return fmt.Errorf("WriteUpdatedMetadataSQL: failed ")
 		}
+		updateSQL.writeQuery(sqlStr)
+	}
+	// Note, the SeriesID stuff was implemented in FW 4.20.14601
+	if kobo.VersionCompare(string(k.fw), "4.20.14601") >= 0 {
+		// Set the SeriesID column correctly
+		// Note, UPDATE FROM is brand spanking new in SQLite 3.33.0 (2020-08-14). We're going to need the latest
+		// client for this one
+		updateSQL.writeQuery(
+			`UPDATE content SET SeriesID = c.SeriesID 
```

Comment:
I'm still not quite certain I've got this right. Initial tests seem promising, but I wouldn't mind another opinion on this.
Note the comment that the latest version of SQLite3 is required for this to work.

Reply:
I think that will work fine, but I'm not too familiar with UPDATE ... FROM myself.